mlir/lib/Driver/CatalystLLVMTarget.cpp

```diff
@@ -39,38 +39,17 @@ void catalyst::driver::registerLLVMTranslations(DialectRegistry &registry)
 
 LogicalResult catalyst::driver::compileObjectFile(const CompilerOptions &options,
                                                   std::shared_ptr<llvm::Module> llvmModule,
+                                                  llvm::TargetMachine *targetMachine,
                                                   StringRef filename)
 {
     using namespace llvm;
 
-    std::string targetTriple = sys::getDefaultTargetTriple();
-
     InitializeAllTargetInfos();
     InitializeAllTargets();
     InitializeAllTargetMCs();
     InitializeAllAsmParsers();
     InitializeAllAsmPrinters();
 
```

Comment:
The fact that removing these triggers a segfault although we already invoked them all just a while earlier in the parent function still puzzles me 🤔  (and these are global functions too)

So does that mean we are "resetting" something here, after initializing them the first time and then modifying some state somewhere? Really weird. 

Any clues @erick-xanadu ?

Reply:
I'm not sure what you are asking here. Moving this from here to earlier fixes the segfault. Maybe I'm missing some context?

The error about the data layout was as following:

1. When O2 and Enzyme was run, the data layout was not added to the LLVM module. In LLVM, the data layout is part of the IR itself but it may be omitted. The default LLVM layout for i64 can be either 4 byte aligned or 8 byte aligned.
2. The GEP canonicalization pass uses the data layout information to decide how much padding is needed in structs. What lead to the discovery of this error was an instruction of the form:
```
%175 = alloca { i1, i64, ptr, ptr }, align 8
%176 = getelementptr inbounds i8, ptr %175, i64 4
```
where we see that the second element (an i64) is aligned to 4 bytes. This is unusual in newer targets. Previously (without the update) it would be this:
```
  %175 = alloca { i1, i64, ptr, ptr }, align 8
  %176 = getelementptr inbounds { i1, i64, ptr, ptr }, ptr %175, i64 0, i32 1
```
which is equivalent only if the second element had a 4 byte alignment. (Which is dependent on the layout)
3. Just before lowering LLVM-IR to binary, the layout was set by this line:
```
    llvmModule->setDataLayout(targetMachine->createDataLayout());
```

which set the alignment for i64s to be 8 byte aligned. This lead to errors where fields were accessed in incorrect locations.

Reply:
This weird thing is related to `test_qnode_exception_dependency` in `async_tests`. 
If I did not abuse Initializations here, the error below would be triggered early and test would not pass.
However, after I merged main into this branch,  the test will pass and leave the error below happening at the very end.

```python
python: /home/tzung-han.juang/Workspace/catalyst-update/mlir/llvm-project/mlir/lib/ExecutionEngine/AsyncRuntime.cpp:53: mlir::runtime::(anonymous namespace)::AsyncRuntime::~AsyncRuntime(): Assertion getNumRefCountedObjects() == 0 && "all ref counted objects must be destroyed"' failed
```

Reply:
I don't think the async runtime error has much to do with the changes above. The error regarding the asynchronous execution happens during the destruction of the async runtime. The async runtime is destroyed when the async runtime library is closed and is no longer referenced, which is beyond our control. (It is part of the OS to decide when).

Reply:
Yes we do not need to duplicate these initializations anymore.